### PR TITLE
Allow passing a type as first argument to `similar`

### LINKED ISSUE #17124: `similar` should accept a type as first parameter
The function `similar` accepts a container as first parameter, but not a container type, in the same way as e.g. `eltype`. (If `dims` is not specified, it should then construct an empty container.)

In some cases, for example in distributed computing, one wants to allocate a container to hold the result of a `map`-like calculation, without having access to the original container. However, the type of the original container is known, as well as the element type of the result. In this case, the interface offered by `similar` is insufficient -- it does not allow allocating a new container without creating an empty dummy container from the type.

A fall-back definition that could be added to Base is
```Julia
similar{CT}(::Type{CT}, specs...) = similar(CT(), specs...)
```
although I assume that many container types will want to provide a more specialized version that avoid creating an empty container.

### PR DESCRIPTION
**Pull request by eschnett**
Closes #17124.

### CODE REVIEW

**Comment by andreasnoack** [comment]
For an array type `T`, why not use the constructor `T(something)` instead of `similar(T, something)`?

**Comment by timholy** [comment]
A perhaps-more-general version is in #17137. That's currently held up due to a weird compile-time performance regression.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -315,6 +315,7 @@ checkbounds(A::AbstractArray) = checkbounds(A, 1) # 0-d case
 typealias SimIdx Union{Integer,UnitRangeInteger}
 """
     similar(array, [element_type=eltype(array)], [dims=size(array)])
+    similar(array_type, [element_type=eltype(array)], [dims=size(array)])
```
Comment (tkelman) [inline base/abstractarray.jl]: should match in the rst

**Comment by eschnett** [comment]
@andreasnoack The use case is `similar(Vector{Int}, Float64)`: all you have is the type `Vector{Int}`, and you want to change the element type to `Float64`.

@timholy Yes, the `allocate_for` there should to do the trick. Let's close this PR here, then.

**Pull request closed without merging**